server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store;
+
+import org.apache.lucene.store.FSDirectory;
+import org.apache.lucene.store.FilterDirectory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.store.Lock;
+import org.opensearch.index.store.remote.file.OnDemandBlockSearchIndexInput;
+import org.opensearch.index.store.remote.filecache.CachedIndexInput;
+import org.opensearch.index.store.remote.filecache.FileCache;
+import org.opensearch.index.store.remote.filecache.FileCachedIndexInput;
+import org.opensearch.index.store.remote.utils.TransferManager;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * a Directory extension that is behaves as a local directory when complete files are present disk,
+ * but fall backs to the block-based on-demand fetch when data is requested that is not present.
+ * The implementation is based on the design: https://github.com/opensearch-project/OpenSearch/issues/7331
+ */
+public class CompositeDirectory extends FilterDirectory {
+    private final FSDirectory localDirectory;
+    private final RemoteSegmentStoreDirectory remoteDirectory;
+    private final TransferManager transferManager;
+    private final FileCache fileCache;
+    /**
+     * Underlying lucene directory for caching
+     */
+    private final FSDirectory localCacheDir;
+    private final FileTrackerImp fileTrackerImp;
+
+    public CompositeDirectory(
+        FSDirectory localDirectory,
+        RemoteSegmentStoreDirectory remoteDirectory,
+        TransferManager transferManager,
+        FileCache fileCache,
+        FSDirectory localCacheDir,
+        FileTrackerImp fileTrackerImp
+    ) {
+        super(localDirectory);
+        this.localDirectory = localDirectory;
+        this.remoteDirectory = remoteDirectory;
+        this.transferManager = transferManager;
+        this.fileCache = fileCache;
+        this.localCacheDir = localCacheDir;
+        this.fileTrackerImp = fileTrackerImp;
+    }
+
+    @Override
+    public String[] listAll() throws IOException {
+        return localDirectory.listAll();
+    }
+
+    @Override
+    public void deleteFile(String name) throws IOException {
+        localDirectory.deleteFile(name);
+    }
+
+    @Override
+    public long fileLength(String name) throws IOException {
+        return localDirectory.fileLength(name);
+    }
+
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) throws IOException {
+        return localDirectory.createOutput(name, context);
```

Comment:
In this case, we rely on the `RemoteStoreRefreshListener` to take care of the sync right?